File: Sully/C/Sully.c
#include <fcntl.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <sys/stat.h>

// The source code of the program itself, stored as a string
char *code = "#include <fcntl.h>%1$c#include <stdlib.h>%1$c#include <stdio.h>%1$c#include <unistd.h>%1$c#include <sys/stat.h>%1$c%1$cchar *code = %3$c%5$s%3$c;%1$c%1$cint		main(void)%1$c{%1$c	char buffer[256];%1$c	int i = %4$d;%1$c%1$c	if (i <= 0)%1$c		return (0);%1$c#ifdef CHILD%1$c	i--;%1$c#endif%1$c	sprintf(buffer, %3$cSully_%2$cd.c%3$c, i);%1$c	chmod(buffer, 0644);%1$c	int fd = open(buffer, O_CREAT | O_TRUNC | O_WRONLY, 0644);%1$c	if (fd < 0)%1$c		return (1);%1$c	dprintf(fd, code, 10, 37, 34, i, code);%1$c	close(fd);%1$c	sprintf(buffer, %3$cgcc -Wall -Wextra -Werror -DCHILD Sully_%2$c1$d.c -o Sully_%2$c1$d%3$c, i);%1$c	system(buffer);%1$c	if (i)%1$c	{%1$c		sprintf(buffer, %3$c./Sully_%2$c1$d%3$c, i);%1$c		system(buffer);%1$c	}%1$c	return (0);%1$c}%1$c";

int main(void)
{
	char buffer[256]; // Buffer to hold file names and commands
	int i = 5; // Initial value of the counter

	// If the counter is less than or equal to 0, exit the program
	if (i <= 0)
		return (0);

#ifdef CHILD
	i--; // Decrement the counter if CHILD is defined
#endif

	// Create the file name for the new source file
	sprintf(buffer, "Sully_%d.c", i);
	chmod(buffer, 0644); // Set the file permissions

	// Open the new source file for writing
	int fd = open(buffer, O_CREAT | O_TRUNC | O_WRONLY, 0644);
	if (fd < 0)
		return (1); // Exit if the file could not be opened

	// Write the source code to the new file
	dprintf(fd, code, 10, 37, 34, i, code);
	close(fd); // Close the file

	// Compile the new source file
	sprintf(buffer, "gcc -Wall -Wextra -Werror -DCHILD Sully_%1$d.c -o Sully_%1$d", i);
	system(buffer);

	// If the counter is still greater than 0, execute the new program
	if (i)
	{
		sprintf(buffer, "./Sully_%1$d", i);
		system(buffer);
	}

	return (0); // Exit the program
}

File: Sully/PERL/Sully.pl
#!/bin/perl

# The source code of the script itself, stored as a string
$source_code = '#!/bin/perl%1$c%1$c$source_code = %2$c%4$s%2$c;%1$c%1$c$counter = %3$d;%1$cif ($counter <= 0) {%1$c    exit;%1$c} elsif ($ENV{CHILD_PROCESS}) {%1$c    $counter--;%1$c}%1$c$script_name = "Sully_$counter.pl";%1$copen(my $file_handle, ">$script_name") or exit;%1$cchmod 0755, $script_name;%1$cprintf $file_handle $source_code, 10, 39, $counter, $source_code;%1$cclose($file_handle);%1$c$ENV{CHILD_PROCESS} = "CHILD_PROCESS";%1$cif ($counter) {%1$c    system("./Sully_$counter.pl");%1$c}%1$c';

# Initialize the counter
$counter = 5;

# Exit if the counter is less than or equal to 0
if ($counter <= 0) {
	exit;
} elsif ($ENV{CHILD_PROCESS}) {
	# Decrement the counter if this is a child process
	$counter--;
}

# Generate the script name based on the counter
$script_name = "Sully_$counter.pl";

# Open a new file with the generated script name for writing
open(my $file_handle, ">$script_name") or exit;

# Set the file permissions to be executable
chmod 0755, $script_name;

# Write the source code to the new file, replacing placeholders with actual values
printf $file_handle $source_code, 10, 39, $counter, $source_code;

# Close the file handle
close($file_handle);

# Set an environment variable to indicate this is a child process
$ENV{CHILD_PROCESS} = "CHILD_PROCESS";

# If the counter is not zero, execute the newly created script
if ($counter) {
	system("./Sully_$counter.pl");
}